### PR TITLE
`schemadiff`: Online DDL support, declarative based

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Addresses https://github.com/vitessio/vitess/issues/16229 and more. Replaces https://github.com/vitessio/vitess/pull/16256 just as a succinct single commit change.

In this PR `schemadiff` takes more ownership in the Online DDL process. Specifically, it takes away much of the logic from `vttablet/onlineddl/vrepl/*` and `vttablet/onlineddl/vrepl.go`, but also:

- Since `schemadiff` is purely declarative and does not use database connections, [query based logic](https://github.com/vitessio/vitess/issues/16229) is replaced with programmatic logic.
- We get rid of old and redundant structures, which we inherited from `gh-ost`: `Column`, `ColumnList`, `UniqueKey` and more. We already have the formal AST types in `go/vt/sqlparser`, and there was a lot of duplication of logic.
- Likewise, we get rid of foreign-key specific logic, of `regexp` based parsing, and more.

Much of the logic removed as mentioned above is reincarnated in `go/vt/schemadiff/onlineddl.go`, where we introduce helper functions specific to Vitess's Online DDL implementations. Intentionally set apart from the general purpose `table.go` or `column.go`, these functions deal with questions such as:

- Given a table change, is any data type _expanded_? This is used for revertibility analysis
- Given a table change, is a new unique constraint introduces? Likewise used for revertibility analysis
- Same for foreign keys.
- Is a given index valid for VReplication iteration?
- Which are the best keys for VReplication iteration?
- Do the before/after tables satisfy iteration key requirements?
- Which columns are shared between before/after tables? Which are dropped? Which iteration keys should we use?

To help formalizing some of the above, we introduce a few data types in `schemadiff` that assist in iteration, filtering, and generally inter-operate with each other to enrich column information:

- `ColumnDefinitionEntity` (introduced in a previous PR) and many helper functions (is this column nullable? Is it numeric? What's the column's charset? etc.)
- `ColumnDefinitionEntityList` - a list of the above `ColumnDefinitionEntity`
- `IndexDefinitionEntity` and many helper functions (is this index unique? Does it have nullable columns? etc.)
- `IndexDefinitionEntityList` - a list of the above `IndexDefinitionEntity`

These inter-connect, as mentioned above. For example, a column `i int` looks to be nullable, but if we know it's part of a `PRIMARY` key, then we implicitly know it is in fact non nullable.

### What this means

This lets `schemadiff` pre-evaluate the ability to run Online DDL on a table, what risks there might be, what the revertibility implications might be, what keys will be used, and so forth, without actually having to read/write anything from/to the the database.

### Testing

Thanks to the declarative-only nature of `schemadiff`, it is easier to do more testing in unit tests. For example, we copy+pasted all the revertible tests from `go/test/endtoend/onlineddl/revert/` and they are now tested in unit tests.
There's otherwise a bunch of tests moved over from `go/vt/vttablet/onlineddl/vrepl`, almost as-is. And otherwise just a bunch more tests. Finally, all of the existing `endtoend` tests, including `vrepl` suite, reverts, scheduler, and more -- all are unchanged and are expected to exhibit no change of behavior.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/16229
- 

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/schemadiff/column.go [hunk 1]
```diff
@@ -181,19 +219,228 @@ func (c *ColumnDefinitionEntity) ColumnDiff(
...
+	}
+	return m
+}
+
+// EnumOrdinalValues returns a map of enum ordinals to their values
+func (c *ColumnDefinitionEntity) EnumOrdinalValues() map[int]string {
+	m := make(map[int]string, len(c.ColumnDefinition.Type.EnumValues))
+	for i, enumValue := range c.ColumnDefinition.Type.EnumValues {
+		// SET and ENUM values are 1 indexed.
+		m[i+1] = enumValue
```
Comment (dbussink) [inline go/vt/schemadiff/column.go:354]: This doesn't seem symmetric with `EnumValuesOrdinals`?

File: go/vt/schemadiff/table.go [hunk 2]
```diff
@@ -445,6 +445,18 @@ type CreateTableEntity struct {
 	Env *Environment
 }
 
+func NewCreateTableEntityFromSQL(env *Environment, sql string) (*CreateTableEntity, error) {
```
Comment (dbussink) [inline go/vt/schemadiff/table.go:448]: @shlomi-noach Do we want to have a symmetric function for views here?

File: go/vt/schemadiff/column.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/column.go:354]: Nice catch. Fixed `EnumValuesOrdinals`, which is as yet unused.

Comment (shlomi-noach) [reply go/vt/schemadiff/column.go:354]: Added tests

File: go/vt/schemadiff/table.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/table.go:448]: Added + testing.

**Review by dbussink — APPROVED** [review]

File: go/vt/schemadiff/key.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		if col.Length != nil {
+			return true
+		}
+	}
+	return false
+}
+
+// ColumnNames returns the names of the columns in the index.
+func (i *IndexDefinitionEntity) ColumnNames() []string {
+	var names []string
```
Comment (mattlord) [inline go/vt/schemadiff/key.go:103]: We could preallocate this slice to a size equal to the length of i.IndexDefinition.Columns.

File: go/vt/schemadiff/onlineddl_test.go [hunk 4]
```diff
@@ -0,0 +1,960 @@
+/*
+Copyright 2022 The Vitess Authors.
```
Comment (mattlord) [inline go/vt/schemadiff/onlineddl_test.go:2]: Should use 2024 here.

File: go/vt/schemadiff/onlineddl.go [hunk 5]
```diff
@@ -0,0 +1,590 @@
+/*
+Copyright 2022 The Vitess Authors.
```
Comment (mattlord) [inline go/vt/schemadiff/onlineddl.go:2]: Should use 2024 here too.

File: go/vt/schemadiff/onlineddl.go [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+	for _, sourceColumn := range sourceColumns.Entities {
+		if sourceColumn.IsGenerated() {
+			continue
+		}
+		isDroppedFromSource := false
+		// Note to a future engineer: you may be tempted to remove this loop based on the
+		// assumption that the later `targetColumn := targetColumns.GetColumn(expectedTargetName)`
+		// check is sufficient. It is not. It is possible that a columns was explicitly dropped
+		// and added (`DROP COLUMN c, ADD COLUMN c INT`) in the same ALTER TABLE statement.
+		// Without checking the ALTER TABLE statement, we would be folled to believe that column
```
Comment (mattlord) [inline go/vt/schemadiff/onlineddl.go:313]: Assuming `folled` should be `fooled`

File: go/vt/vttablet/onlineddl/vrepl_test.go [hunk 7]
```diff
@@ -4,3 +4,224 @@
 */
```
Comment (mattlord) [inline go/vt/vttablet/onlineddl/vrepl_test.go:4]: Can we remove this copyright at this point?

Either way, we should also add the vitess copyright.

**Review by mattlord — APPROVED** [review]

File: go/vt/schemadiff/key.go (hunk 3 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/key.go]: fixed.

File: go/vt/schemadiff/onlineddl.go (hunk 5 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/onlineddl.go:2]: fixed.

File: go/vt/schemadiff/onlineddl.go (hunk 6 above)
Comment (shlomi-noach) [reply go/vt/schemadiff/onlineddl.go]: fixed.

File: go/vt/vttablet/onlineddl/vrepl_test.go (hunk 7 above)
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/vrepl_test.go:15]: Good catch. For this file the copyright should be just vitess. There is not derivative work from `freno` here.

Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/vrepl_test.go:15]: fixed.

**Pull request merged**